morgoth/addons/models.py

```diff
@@ -1,13 +1,94 @@
+import hashlib
+import os
+
+from urllib import request
+
 from django.db import models
 
+from dirtyfields import DirtyFieldsMixin
+
+
+PLATFORMS = (
+    'Darwin_x86-gcc3-u-i386-x86_64',
+    'Darwin_x86_64-gcc3-u-i386-x86_64',
+    'Linux_x86-gcc3',
+    'Linux_x86_64-gcc3',
+    'WINNT_x86-msvc',
+    'WINNT_x86-msvc-x64',
+    'WINNT_x86-msvc-x86',
+    'WINNT_x86_64-msvc',
+    'WINNT_x86_64-msvc-x64',
+)
+
 
-class Addon(models.Model):
+class Addon(DirtyFieldsMixin, models.Model):
     name = models.CharField(max_length=100)
     version = models.CharField(max_length=32)
-    ftp_url = models.URLField()
+    ftp_url = models.URLField(blank=True, null=True)
+    xpi_hash = models.CharField(max_length=128, null=True)
+    xpi_filesize = models.IntegerField(null=True)
+
+    def save(self, *args, **kwargs):
+        if self.is_dirty():
+            dirty_fields = self.get_dirty_fields()
+
+            if 'ftp_url' in dirty_fields and self.ftp_url:
+                # Download the XPI file
+                file_name, headers = request.urlretrieve(self.ftp_url)
```

Comment:
i could potentially do a `HEAD` request but that doesn't help with hashing the XPI.